src/simulation/Simulation.cpp

```diff
@@ -672,4 +672,43 @@ Simulation::metricsSummary(string domain)
     }
     return out.str();
 }
+
+void
+Simulation::sendMessageAndCrank(StellarMessage& msg, NodeID initiator,
+                                NodeID acceptor)
+{
+    auto it = std::find_if(
+        std::begin(mLoopbackConnections), std::end(mLoopbackConnections),
+        [&](std::shared_ptr<LoopbackPeerConnection> const& conn) {
+            return conn->getInitiator()
+                           ->getApp()
+                           .getConfig()
+                           .NODE_SEED.getPublicKey() == initiator &&
+                   conn->getAcceptor()
+                           ->getApp()
+                           .getConfig()
+                           .NODE_SEED.getPublicKey() == acceptor;
+        });
+    if (it != std::end(mLoopbackConnections))
+    {
+        auto initiator = (*it)->getInitiator();
+        auto acceptor = (*it)->getAcceptor();
+
+        initiator->getApp().getClock().postToCurrentCrank(
+            [initiator, msg]() { initiator->Peer::sendMessage(msg); });
+
+        crankForAtMost(std::chrono::milliseconds{500}, false);
+
+        // clear all queues and cancel all events
+        initiator->clearInAndOutQueues();
+        acceptor->clearInAndOutQueues();
+
+        while (initiator->getApp().getClock().cancelAllEvents())
+            ;
+        while (acceptor->getApp().getClock().cancelAllEvents())
+            ;
+        while (mClock.cancelAllEvents())
```

Comment:
why do you need to call `cancelAllEvents` on `mClock`?

Reply:
Removed. Originally, I believed this to be a source of non-determinism -- events persisting between executions -- however, now that I have looked closer it looks like the issue was only with `LoopbackPeer` in/out message queues.

Reply:
I do not, remove -- double checked and events only persist on `acceptor` and `initiator`, not the idle app.